[DOC] Update to security chapter for pkcs #12 (#2310)
strimzi/strimzi-kafka-operator: documentation/modules/security/proc-configuring-external-clients-to-trust-cluster-ca.adoc

// Module included in the following assemblies:
//
// assembly-security.adoc

[id='configuring-external-clients-to-trust-cluster-ca-{context}']
= Configuring external clients to trust the cluster CA

This procedure describes how to configure a Kafka client that resides outside the Kubernetes cluster – connecting to the `external` listener on port 9094 – to trust the cluster CA certificate.
Follow this procedure when setting up the client and during the renewal period, when the old clients CA certificate is replaced.

> Follow the steps to configure trust certificates signed by the cluster CA for Java-based Kafka Producer, Consumer, and Streams APIs.

Reviewer: ```suggestion
Follow the steps to configure trust certificates that are signed by the cluster CA for Java-based Kafka Producer, Consumer, and Streams APIs.
```

> Follow the steps to configure trust certificates signed by the cluster CA for Java-based Kafka Producer, Consumer, and Streams APIs.

Reviewer: ```suggestion
Follow the steps to configure trust certificates signed by the cluster CA for Java-based Kafka Producer, Consumer, and Kafka Streams APIs.
```

> Choose the steps according to the certificate format for the cluster CA: PKCS#12 (`.p12`) or PEM (`.crt`).

Reviewer: ```suggestion
Choose the steps to follow according to the certificate format of the cluster CA: PKCS #12 (`.p12`) or PEM (`.crt`).
```

> IMPORTANT: The `_<cluster-name>_-cluster-ca-cert` `Secret` will contain more than one CA certificate during CA certificate renewal.

Reviewer: Is "during CA certificate renewal" concrete enough?

"During the CA certificate renewal period"?

Clients must add _all_ of them to their truststores.

.Prerequisites

* The Cluster Operator is running.
* A `Kafka` resource within the Kubernetes cluster.

> * A Kafka client application outside the Kubernetes cluster which will connect using TLS and needs to trust the cluster CA certificate.

Reviewer: ```suggestion
* A Kafka client application outside the Kubernetes cluster that will connect using TLS and needs to trust the cluster CA certificate.
```

.Using PKCS #12 format (.p12)

. Extract the cluster CA certificate and password from the generated `_<cluster-name>_-cluster-ca-cert` Secret.

> +

Reviewer: What do you think about adding explanatory comments to these commands?

Author: If you think it's adds clarity. So what I've done, so I don't repeat the same thing in both mini procedures, is I've added this to the intro to the procedure:

> The steps describe how to:
> 
> - Obtain the certificate that verifies the identity of the Kafka cluster
> - Configure the client credentials to use the extracted certificate

[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.p12}' | base64 -d > ca.p12
+
[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.password}' | base64 -d > ca.password

. Configure the Kafka client with the following properties:
+

> * Security protocol option:

Reviewer: ```suggestion
* The security protocol option:
```

> ** `security.protocol: SSL` when using TLS for encryption (with or without TLS authentication)

Reviewer: ```suggestion
** `security.protocol: SSL` when using TLS for encryption (with or without TLS authentication).
```

** `security.protocol: SASL_SSL` when using SCRAM-SHA authentication over TLS.
* `ssl.truststore.location`: the truststore location where the certificates were imported.
* `ssl.truststore.password`: the password for accessing the truststore. This property can be omitted if it is not needed by the truststore.

.Using PEM format (.crt)

. Extract the cluster CA certificate from the generated `_<cluster-name>_-cluster-ca-cert` Secret.
+
[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.crt}' | base64 -d > ca.crt

. Import the CA certificates into the JVM's truststore using the following `keytool` command:
+
[source,shell]
keytool -keystore client.truststore.jks -alias CARoot -import -file ca.crt

. Configure the Kafka client with the following properties:
+

> * Security protocol option:

Reviewer: ```suggestion
* The security protocol option:
```

> ** `security.protocol: SSL` when using TLS for encryption (with or without TLS authentication)

Reviewer: ```suggestion
** `security.protocol: SSL` when using TLS for encryption (with or without TLS authentication).
```

// Module included in the following assemblies:
//
// assembly-security.adoc

[id='configuring-external-clients-to-trust-cluster-ca-{context}']
= Configuring external clients to trust the cluster CA

This procedure describes how to configure a Kafka client that resides outside the Kubernetes cluster – connecting to the `external` listener on port 9094 – to trust the cluster CA certificate.
Follow this procedure when setting up the client and during the renewal period, when the old clients CA certificate is replaced.

Follow the steps to configure trust certificates that are signed by the cluster CA for Java-based Kafka Producer, Consumer, and Streams APIs.

Choose the steps to follow according to the certificate format of the cluster CA: PKCS #12 (`.p12`) or PEM (`.crt`).

The steps describe how to:

* Obtain the certificate from the Cluster Secret that verifies the identity of the Kafka cluster

> * Configure the client credentials to use the extracted certificate

Author: I removed the bullet here and the related one in the internal procedure.

+
[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.p12}' | base64 -d > ca.p12
+
[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.password}' | base64 -d > ca.password

. Configure the Kafka client with the following properties:
+
* A security protocol option:
** `security.protocol: SSL` when using TLS for encryption (with or without TLS authentication).
** `security.protocol: SASL_SSL` when using SCRAM-SHA authentication over TLS.
* `ssl.truststore.location`: the truststore location where the certificates were imported.
* `ssl.truststore.password`: the password for accessing the truststore. This property can be omitted if it is not needed by the truststore.

.Using PEM format (.crt)

. Extract the cluster CA certificate from the generated `_<cluster-name>_-cluster-ca-cert` Secret.
+
[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.crt}' | base64 -d > ca.crt

. Import the CA certificates into the JVM's truststore using the following `keytool` command:
+
[source,shell]

> keytool -keystore client.truststore.jks -alias CARoot -import -file ca.crt

Author: Thanks @scholzj. @ppatierno - I'm going to reduce the steps for the PEM procedure.

> keytool -keystore client.truststore.jks -alias CARoot -import -file ca.crt

Author: .Using PEM format (.crt)

. Extract the cluster CA certificate from the generated `_<cluster-name>_-cluster-ca-cert` Secret.

[source,shell,subs="+quotes"]
kubectl get secret _<cluster-name>_-cluster-ca-cert -o jsonpath='{.data.ca\.crt}' | base64 -d > ca.crt

. Use the certificate with clients that use certificates in X.509 format.

> . Configure the Kafka client with the following properties:

Author: Removed. see previous comment
